# Deployment und Abschluss - pycaret

## Install

ref: https://pycaret.gitbook.io/docs/get-started/installation
* You can install PyCaret with Python's pip package manager:

## Load and prep Data

In [4]:
## load data
import numpy as np
import pandas as pd

datapath = '../3_data'
from os import chdir; chdir(datapath)
dataset = pd.read_csv('bank_data_prep.csv')
print(dataset.shape)

(9860, 30)


In [5]:
## remove duration
dataset = dataset.drop("duration", axis = 1)

In [6]:
## train - test - split
from sklearn.model_selection import train_test_split
data, data_unseen = train_test_split(dataset, train_size=0.9, random_state=1234)

print('Data for Modeling: ' + str(data.shape))
print('Unseen Data For Predictions: ' + str(data_unseen.shape))

Data for Modeling: (8874, 29)
Unseen Data For Predictions: (986, 29)


## Run a Classication Experiment

### Init setup

In [9]:
from pycaret.classification import *
s = setup(
    data = data, 
    target = 'y', 
    fold = 5, ## defaul = 10
    session_id=1234) ## random seed

### Show available Models (for Classification)

In [11]:
models()

Name  \
ID                                          
lr                    Logistic Regression   
knn                K Neighbors Classifier   
nb                            Naive Bayes   
dt               Decision Tree Classifier   
svm                   SVM - Linear Kernel   
rbfsvm                SVM - Radial Kernel   
gpc           Gaussian Process Classifier   
mlp                        MLP Classifier   
ridge                    Ridge Classifier   
rf               Random Forest Classifier   
qda       Quadratic Discriminant Analysis   
ada                  Ada Boost Classifier   
gbc          Gradient Boosting Classifier   
lda          Linear Discriminant Analysis   
et                 Extra Trees Classifier   
lightgbm  Light Gradient Boosting Machine   
catboost              CatBoost Classifier   
dummy                    Dummy Classifier   

                                                  Reference  Turbo  
ID                                                                  
lr        sklearn.linear_model._logistic.LogisticRegression   True  
knn       sklearn.neighbors._classification.KNeighborsCl...   True  
nb                           sklearn.naive_bayes.GaussianNB   True  
dt             sklearn.tree._classes.DecisionTreeClassifier   True  
svm       sklearn.linear_model._stochastic_gradient.SGDC...   True  
rbfsvm                             sklearn.svm._classes.SVC  False  
gpc       sklearn.gaussian_process._gpc.GaussianProcessC...  False  
mlp       sklearn.neural_network._multilayer_perceptron....  False  
ridge           sklearn.linear_model._ridge.RidgeClassifier   True  
rf          sklearn.ensemble._forest.RandomForestClassifier   True  
qda       sklearn.discriminant_analysis.QuadraticDiscrim...   True  
ada       sklearn.ensemble._weight_boosting.AdaBoostClas...   True  
gbc         sklearn.ensemble._gb.GradientBoostingClassifier   True  
lda       sklearn.discriminant_analysis.LinearDiscrimina...   True  
et            sklearn.ensemble._forest.ExtraTreesClassifier   True  
lightgbm                    lightgbm.sklearn.LGBMClassifier   True  
catboost                   catboost.core.CatBoostClassifier   True  
dummy                         sklearn.dummy.DummyClassifier   True

### Model Training and Selection

In [13]:
best_model = compare_models()

Processing:   0%|          | 0/65 [00:00<?, ?it/s]

MCC, see: https://en.wikipedia.org/wiki/Phi_coefficient

In [15]:
print(best_model)

GradientBoostingClassifier(ccp_alpha=0.0, criterion='friedman_mse', init=None,
                           learning_rate=0.1, loss='log_loss', max_depth=3,
                           max_features=None, max_leaf_nodes=None,
                           min_impurity_decrease=0.0, min_samples_leaf=1,
                           min_samples_split=2, min_weight_fraction_leaf=0.0,
                           n_estimators=100, n_iter_no_change=None,
                           random_state=1234, subsample=1.0, tol=0.0001,
                           validation_fraction=0.1, verbose=0,
                           warm_start=False)


### Analyze best Model

In [17]:
evaluate_model(best_model)

interactive(children=(ToggleButtons(description='Plot Type:', icons=('',), options=(('Pipeline Plot', 'pipelin…

### Predict on unseen Data

In [19]:
predictions = predict_model(best_model, data=data_unseen)
print(predictions.head())

       age  education  housing  contact_cellular  month  day_of_week  \
2809  36.0          6        0                 1      5            2   
4052  34.0          6        1                 1      8            3   
658   36.0          6        1                 0      5            2   
786   40.0          5        0                 0      5            3   
6675  36.0          6        0                 1     11            3   

      campaign  pdays  previous  emp_var_rate  ...  job_student  \
2809  0.477121      0         0          -1.8  ...        False   
4052  0.698970      0         0           1.4  ...        False   
658   0.903090      0         0          -1.8  ...        False   
786   0.698970      0         0           1.1  ...        False   
6675  0.477121      0         0          -0.1  ...        False   

      job_technician  job_unemployed  marital_married  marital_single  \
2809           False           False            False            True   
4052            Tr

### Save best Model Pipeline

## Tune a specific Model
ref: https://pycaret.gitbook.io/docs/get-started/functions/optimize

### rf:  Random Forest Classifier

In [23]:
## create model
model_rf = create_model('rf')

Processing:   0%|          | 0/4 [00:00<?, ?it/s]

In [24]:
## tune model
model_rf_tuned = tune_model(model_rf)

Processing:   0%|          | 0/7 [00:00<?, ?it/s]

Fitting 5 folds for each of 10 candidates, totalling 50 fits


In [25]:
## parameters of default model
print(model_rf)

RandomForestClassifier(bootstrap=True, ccp_alpha=0.0, class_weight=None,
                       criterion='gini', max_depth=None, max_features='sqrt',
                       max_leaf_nodes=None, max_samples=None,
                       min_impurity_decrease=0.0, min_samples_leaf=1,
                       min_samples_split=2, min_weight_fraction_leaf=0.0,
                       monotonic_cst=None, n_estimators=100, n_jobs=-1,
                       oob_score=False, random_state=1234, verbose=0,
                       warm_start=False)


In [26]:
## parameters of tuned model
print(model_rf_tuned)

RandomForestClassifier(bootstrap=True, ccp_alpha=0.0,
                       class_weight='balanced_subsample', criterion='gini',
                       max_depth=10, max_features='sqrt', max_leaf_nodes=None,
                       max_samples=None, min_impurity_decrease=0,
                       min_samples_leaf=5, min_samples_split=7,
                       min_weight_fraction_leaf=0.0, monotonic_cst=None,
                       n_estimators=160, n_jobs=-1, oob_score=False,
                       random_state=1234, verbose=0, warm_start=False)


comparison of the default parameters and the tuned parameters

| parameter | default | tuned |
| :--- | :--- | :--- |
| ccp_alpha | 0 | 0 |
| class_weight | None | 'balanced_subsample' |
| criterion | 'gini' | 'gini' |
| max_depth | None | 10 |
| max_features | 'sqrt' | 'sqrt' |
| max_leaf_nodes | None | None |
| max_samples | None | None |
| min_impurity_decrease | 0 | 0 |
| min_samples_leaf | 1 | 5 |
| min_samples_split | 2 | 7 |
| min_weight_fraction_leaf | 0 | 0 |
| monotonic_cst | None | None |
| n_estimators | 100 | 160 |
| n_jobs | -1 | -1 |
| oob_score | False | False |
| random_state | 1234 | 1234 |
| verbose | 0 | 0 |
| warm_start | False | False |